network-ntt/src/server.rs

```diff
@@ -0,0 +1,313 @@
+//! on the specified port and client connects to the remote service.
+//! Once node node establised a connection to the other one, the other
+//! once can build an opposite connection and nodes can talk to
+//! each other.
+//!
+//! ```text
+//!                                     +---- server: sock1
+//!                                     |      |
+//!                                     |      |-- connection1
+//!                                     |      |-- connection2
+//!  +--------------+     +----------+  |
+//!  | network-core | ----| NT state |--+---- server: sock2
+//!  +--------------+     +----------+  |      |
+//!                                     |      |-- connection3
+//!                                     |
+//!                                     +---- client: remote-sock1
+//! ```
+//!
+//! As NT state is shared all the services could provide a uniform
+//! access to the chain state. And one could build very flexible
+//! topology.
+//use crate::reply_handle::ReplyHandle;
+//use chain_core::property::{Deserialize, Serialize};
+use futures::prelude::*;
+use futures::{future, stream::{self,Stream}, sync::mpsc};
+use network_core::server;
+use network_core::server::block::BlockService;
+use network_core::server::transaction::TransactionService;
+use std::net::SocketAddr;
+use tokio::net::{TcpListener, TcpStream};
+use tokio_protocol::{protocol, Connection, Inbound, Message};
+
+// These types are loaded from cardano package,
+// because tokio-protocol is not generalized yet.
+// This should be fixed and then these types should go.
+use cardano::{
+    block::{HeaderHash},
+    tx::{TxAux},
+};
+
+/// Network transport server configuration.
+pub struct Config {
+    listen: Vec<SocketAddr>,
+    connect: Vec<SocketAddr>,
+}
+
+/// Internal structure of network transport node.
+#[derive(Clone)]
+pub struct Server<N> {
+    node: N,
+}
+
+/// Run NTT server.
+pub fn run<N: 'static>(config: Config, node: N)
+where
+    N: std::marker::Sync + std::marker::Send + server::Node + Clone + Copy,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::TipFuture: std::marker::Send,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::GetHeadersFuture: std::marker::Send,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::StreamBlocksToTipFuture: std::marker::Send,
+    <<N as network_core::server::Node>::TransactionService as network_core::server::transaction::TransactionService>::ProposeTransactionsFuture: std::marker::Send,
+{
+
+        let listener =
+            stream::iter_ok(config.listen)
+                .for_each(move |sockaddr| {
+                    let server = Server {
+                          node: node.clone(),
+                    };
+                    run_listen_socket(sockaddr, server) 
+                });
+        let connections =
```

Comment:
There should be a `Connection` object to encapsulate peer connection state, like in `network-grpc`

Reply:
Done